quinn-proto/src/connection/mod.rs

```diff
@@ -389,76 +405,160 @@ where
             }
         }
 
-        // Select the set of spaces that have data to send so we can try to coalesce them
-        let (spaces, close) = match self.state {
+        // Check whether we need to send a close message
+        let mut close = match self.state {
             State::Drained => {
                 self.app_limited = true;
                 return None;
             }
             State::Draining | State::Closed(_) => {
+                // TODO: This seems to lose the `close` flag if it can't be
+                // immediately enqueued
                 if mem::replace(&mut self.close, false) {
-                    (vec![self.highest_space], true)
+                    true
                 } else {
                     self.app_limited = true;
                     return None;
                 }
             }
-            _ => (
-                SpaceId::iter()
-                    .filter(|&x| {
-                        (self.spaces[x].crypto.is_some() && self.spaces[x].can_send())
-                            || (x == SpaceId::Data
-                                && ((self.spaces[x].crypto.is_some() && self.can_send_1rtt())
-                                    || (self.zero_rtt_crypto.is_some()
-                                        && self.side.is_client()
-                                        && (self.spaces[x].can_send() || self.can_send_1rtt()))))
-                    })
-                    .collect(),
-                false,
-            ),
+            _ => false,
         };
 
-        let mut buf = Vec::with_capacity(self.path.mtu as usize);
+        let mut buf = Vec::new();
         // Reserving capacity can provide more capacity than we asked for.
         // However we are not allowed to write more than MTU size. Therefore
         // the maximum capacity is tracked separately.
-        let buf_capacity = self.path.mtu as usize;
-
-        let mut coalesce = spaces.len() > 1;
-        let pad_space = spaces.last().cloned().filter(|_| {
-            self.side.is_client() && spaces.first() == Some(&SpaceId::Initial)
-                || self.path.challenge.is_some()
-                || self.path_response.is_some()
-        });
+        let mut buf_capacity = 0;
 
+        let mut coalesce = true;
+        let mut builder: Option<PacketBuilder> = None;
+        let mut sent_frames = None;
+        let mut min_datagram_size = 0;
```

Comment:
Could we replace this with `let mut pad_datagram = false`? It seems to only have two possible values, and this would let us avoid writing the nonzero case twice.

Reply:
It has 3 possible values if we take the next commit into account:
1. Don't pad
2. Pad to min initial size
3. Pad to MTU size

is it worth it to make it an enum?

Reply:
And I agree, the calculations being spread out over a variety of places isn't great

Reply:
Where's option 3 arise? I only see two assignments to nonzero values, and both are exactly `min_datagram_size.max(builder.datagram_start + MIN_INITIAL_SIZE - builder.tag_len)`

Reply:
Oh, are you referring to a commit that hasn't been pushed yet? If so I'll take your word for it and leave as-is for now; an enum seems questionable.

Reply:
Here's a proposal:

Have a
```
enum PadDatagram {
   DontPad,
   PadToMinInitialSize,
   PadToMtuSize
}
```

use here `let mut pad_datagram = PadDatagram::DontPad`

and add a `PacketBuilder::set_padding(&mut self, PadDatagram, mtu)` which gets called instead of manually fiddling with `builder.min_size`

Reply:
> Oh, are you referring to a commit that hasn't been pushed yet? If so I'll take your word for it and leave as-is for now; an enum seems questionable.

Oh sorry, it's confusing. It is pushed as the last GSO commit. But there it doesn't set `min_datagram_size` and copies it over. We directly set `builder.min_size = builder.datagram_start + (self.path.mtu as usize) - builder.tag_len;` instead.

Reply:
Right, so in that case can we get away with a `bool` here?

Reply:
Yes, I updated with a `bool` and 2 helper methods for `PacketBuilder`.
I now think the final padding instruction could be passed into `finish_and_track_packet`. But that would require the enum again.

Reply:
So probably its just fine as is.